runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/WorkerPoolTest.java

```diff
@@ -50,22 +81,38 @@ protected IStatus run(IProgressMonitor monitor) {
 		}
 
 		// wait for jobs to reach the barrier
-		barrier.await(10, TimeUnit.SECONDS);
+		parallelJobStartBarrier.await(BARRIER_TIMEOUT_IN_SEC, TimeUnit.SECONDS);
+
+		long startTimeInMSec = System.currentTimeMillis();
 
-		// this is the ugly part, wait until worker threads become idle
-		Thread.sleep(5 * 1000L);
+		// wait for workerThreadCount to sink below MAX_ALLOWED_IDLE_WORKER_THREADS
+		while (true) {
+			int workerThreadCount = getWorkerThreadCount();
+
+			if (workerThreadCount <= MAX_ALLOWED_IDLE_WORKER_THREADS) {
+				break; // Exit loop when the condition is met
+			}
 
-		// count worker threads, must be less than WorkerPool.MAX_THREADS
+			if (System.currentTimeMillis() - startTimeInMSec > WORKER_TIMEOUT_IN_MSEC) {
+				Assert.fail("Timeout reached! Too many worker threads active: " + workerThreadCount + ", expected <= "
+						+ MAX_ALLOWED_IDLE_WORKER_THREADS);
+			}
+
+			Thread.yield();
+		}
```

Comment:
It's a matter of taste, but I'd prefer to make the loop condition explicit in the loop header rather than having a break somewhere in the loop body, i.e. have:
`while(workerThreadCount > MAX_ALLOWED_IDLE_WORKER_THREADS)`
and define the variable before the loop and overwrite it at the end of each iteration with the current value.